### ISSUE TITLE
What are the requirements of the model by using these visual code?

### ISSUE THREAD

**Issue by long123524** [frgfm/torch-cam#171]
### Bug description

![image](https://user-images.githubusercontent.com/73996375/178204623-e411b303-a57f-46df-893f-c00544f09055.png)

I have got a error, we failed to use these method to obtain a results. I have got  4D matrix, but it needs to be 2D. My model have not a fully connected layer，while consists of many convolution layers.  Should I how to deal with this problem? Thank you

### Code snippet to reproduce the bug

    def _backprop(self, scores: Tensor, class_idx: int) -> None:
        """Backpropagate the loss for a specific output class"""

        # Backpropagate to get the gradients on the hooked layer

        loss = scores[:, class_idx].sum()
        self.model.zero_grad()
        loss.backward(retain_graph=True)

### Error traceback

![image](https://user-images.githubusercontent.com/73996375/178206024-032493f0-e909-4b0e-aad7-0c48d2dc0877.png)

### Environment

Cuda 11.0
torch-gpu==1.7.0

**Comment by frgfm** [comment]
Hello @long123524 :wave: 

Sorry about this! But I'll need more information for understandable reasons :sweat_smile: 
Your snippet is not runnable, could you share a minimal snippet (including imports) that would reproduce this bug please?

According to your console output , it looks like you used a UNet. Are you tried to get the CAM of a segmentation model? :thinking:

**Comment by long123524** [comment]
You are right. I use a encoder-decoder architecture segmentation model, like UNet. Should I how to obtain a cam? Looking forward to your reply.

**Comment by frgfm** [comment]
Alright, so this is a much wider discussion then!

I suggest checking https://github.com/frgfm/torch-cam/discussions/150 :)
CAMs were designed to provide spatial influence at a given layer about a scalar output: 
- this scalar output can be the probability of a given class of a multi-class classification model, a binary classification model output, a video classification model. We want a single channel spatial representation meaning (H, W) shape for the CAM, you can simplify this to consider this as the gradient of the activation at a given layer relative to this scalar output.
- now the problem is that with segmentation models, you get (C, H', W') outputs for 2D models and not (C,). So the "CAM" of a segmentation model, following the previous comparison, would be the partial derivative of the activation at a given layer, relative to both the value & location of the scalar. This means that this "CAM" would not be of shape (H, W) like previously but (H, W, H', W'), which is not exactly fit for visualization.

To make this clearer:
- in the first case, you check for every spatial location in the activation map, its influence on the scalar output
- in the second case, that would be checking for every spatial location in the activation map, its influence on the 2D output

So, may I ask: what are you exactly trying to achieve with this CAM? :)

**Comment by long123524** [comment]
I have the same problem as https://github.com/frgfm/torch-cam/discussions/150. The aim of our work to obtain a CAM of semantic segmentation model, like UNet. However, I fails to get it.

**Comment by frgfm** [comment]
@long123524 the whole point of #150 is about the relevance of CAM for semantic segmentation

Again, there is no definition for that :sweat_smile: 
So could you elaborate on "obtain a CAM of semantic segmentation model" please? :pray:

**Comment by Harry-KIT** [comment]
Hi @frgfm. Good job!
I wanted to know how you generated video_example_wallaby.gif using using some code.

**Comment by frgfm** [comment]
@Harry-KIT this is a different topic than this issue but briefly: reading through the frames of a video with OpenCV, classifying the images with the model, computing the CAM, then producing the GIF with the resulting frames :ok_hand:

**Comment by Harry-KIT** [comment]
https://user-images.githubusercontent.com/57295863/182058122-5a53fb27-3e8b-4ba8-97ff-5183d1f2dbb5.mp4

Hi @frgfm. Thanks for your help!

**Comment by frgfm** [comment]
Back to the main topic of the issue :)
@long123524 I think I'll close this issue if I understood you correctly because there is a need for a definition of a CAM for a semantic segmentation model. So apart from if I got everything wrong, your request is about something that doesn't have a mathematical definition yet :sweat_smile:

**Comment by frgfm** [comment]
Closing this as per my previous comment
